lib/install/management/inventory.go

```diff
@@ -0,0 +1,217 @@
+	}, tasks.IsTransientError)
+	if err != nil {
+		d.op.Errorf("Failure moving VCH(%s) into folder(%s): %s", name, tempName, err)
+		return err
+	}
+	d.op.Debugf("%s moved into folder", name)
+	//now rename the temp folder
+	d.renameFolder(target, name)
+	return nil
+}
+
+// renameFolder renames the the provided folder to the specified name
+func (d *Dispatcher) renameFolder(target *object.Folder, name string) error {
+	targetOriginal := target.Name()
+	if targetOriginal == name {
+		//noop
+		return nil
+	}
+
+	_, err := tasks.WaitForResultAndRetryIf(d.op, func(op context.Context) (tasks.Task, error) {
+		return target.Rename(d.op, name)
+	}, tasks.IsTransientError)
+	if err != nil {
+		d.op.Errorf("Unable to rename folder(%s) to %s during inventory update: %s", targetOriginal, name, err)
+		return err
+	}
+	d.op.Debugf("Folder renamed from %s to %s", targetOriginal, name)
+	return nil
+}
+
+// folderCleanup searches for folders with a name resembling a temporary folder during
+// inventoryUpdate.  This will remove empty folders that were not deleted during a failed upgrade.
+func (d *Dispatcher) folderCleanup(name string, tempPattern string) {
+	// check for any orphaned upgrade folders
+	folders, err := d.session.Finder.FolderList(d.op, path.Join(d.session.VMFolder.InventoryPath, "*"))
+	if err != nil {
+		d.op.Debugf("Error searching for orphaned upgrade folders: %s", err)
+		return
+	}
+	d.op.Debugf("Searching %d folders for partial name(%s) match", len(folders), tempPattern)
+	for _, f := range folders {
+		if strings.Contains(f.Name(), tempPattern) {
+			d.op.Debugf("Found Orphaned Folder: %s", f.Name())
+			// delete if it's not empty
+			d.deleteFolder(f)
+		}
+	}
+}
+
+// containerRefs returns an array of ManagedObjectRefrences that are the containerVMs for the pool
+func (d *Dispatcher) containerRefs(pool *object.ResourcePool) ([]types.ManagedObjectReference, error) {
+	defer trace.End(trace.Begin(pool.InventoryPath, d.op))
+	// slice for validated containers
+	var containers []types.ManagedObjectReference
+	// find the vms for this resource pool
+	vms, err := d.children(pool)
+	if err != nil {
+		return nil, err
+	}
+	// determine which are containers
+	for _, vm := range vms {
+		if d.isContainer(vm) {
+			d.op.Debugf("Found Container: %s", vm.Summary.Config.Name)
+			containers = append(containers, vm.Reference())
+		}
+	}
+	d.op.Debugf("Found %d containers", len(containers))
+	return containers, nil
+}
+
+// children returns an array of VirtualMachines with properties populated
+func (d *Dispatcher) children(pool *object.ResourcePool) ([]mo.VirtualMachine, error) {
+	defer trace.End(trace.Begin(pool.InventoryPath, d.op))
+
+	// Get the vmRefs for this resource pool
+	refs, err := compute.VM(d.op, d.session, pool)
+	if err != nil {
+		err = fmt.Errorf("Error retrieving vms: %s", err)
+		return nil, err
+	}
+
+	return vm.Attributes(d.op, d.session, refs)
+}
+
+func (d *Dispatcher) isContainer(vm mo.VirtualMachine) bool {
```

Comment:
How does this relate to `isContainerVM`?

https://github.com/vmware/vic/blob/0ff435cd277c44b90b36f93b837c1c45d47500b3/lib/install/management/appliance.go#L95-L111

Reply:
It will eventually replace it...instead of doing that in this PR (that's already huge) I was waiting to break it out into chunks -- I planned on documenting what I think we should do and then discuss with you, George, etc..

Basically we need to rework the interactions w/VC as vic-machine is very chatty...for example when you delete it will pull the config 2x for each VM...I determined the other day on a `vic-machine delete` there are roughly 9 calls to VC for each container it's deleting...so have 10 containers that's 90 calls...and that's just a portion of the interaction...we need to do a much better job of protecting the infrastructure...